presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentStreamingPageSource.java

```diff
@@ -0,0 +1,135 @@
+    {
+        if (closed) {
+            return null;
+        }
+        if (serverResponseIterator == null) {
+            serverResponseIterator = queryPinot(split);
+        }
+        ByteBuffer byteBuffer = null;
+        if (serverResponseIterator.hasNext()) {
+            long startTimeNanos = System.nanoTime();
+            ServerResponse serverResponse = serverResponseIterator.next();
+            readTimeNanos += System.nanoTime() - startTimeNanos;
+            switch (serverResponse.getResponseType()) {
+                case Constants.Response.ResponseType.DATA:
+                    estimatedMemoryUsageInBytes = serverResponse.getSerializedSize();
+                    // Store each dataTable which will later be constructed into Pages.
+                    try {
+                        byteBuffer = serverResponse.getPayloadReadOnlyByteBuffer();
+                        DataTable dataTable = serverResponse.getDataTable(byteBuffer);
+                        checkExceptions(dataTable, split, PinotSessionProperties.isMarkDataFetchExceptionsAsRetriable(session));
+                        currentDataTable = new PinotSegmentPageSource.PinotDataTableWithSize(dataTable, serverResponse.getSerializedSize());
+                    }
+                    catch (IOException e) {
+                        throw new PinotException(
+                            PINOT_DATA_FETCH_EXCEPTION,
+                            split.getSegmentPinotQuery(),
+                            String.format("Encountered Pinot exceptions when fetching data table from Split: < %s >", split), e);
+                    }
+                    break;
+                case Constants.Response.ResponseType.METADATA:
+                    // The last part of the response is Metadata
+                    currentDataTable = null;
+                    serverResponseIterator = null;
+                    close();
+                    return null;
+                default:
+                    throw new PinotException(
+                        PINOT_UNEXPECTED_RESPONSE,
+                        split.getSegmentPinotQuery(),
+                        String.format("Encountered Pinot exceptions, unknown response type - %s", serverResponse.getResponseType()));
+            }
+        }
+        Page page = fillNextPage();
+        completedPositions += currentDataTable.getDataTable().getNumberOfRows();
+        if (byteBuffer != null) {
+            byteBuffer.clear();
+        }
+        return page;
+    }
+
+    private Iterator<ServerResponse> queryPinot(PinotSplit split)
+    {
+        String sql = split.getSegmentPinotQuery().orElseThrow(() -> new PinotException(PINOT_INVALID_SEGMENT_QUERY_GENERATED, Optional.empty(), "Expected the segment split to contain the pinot query"));
+        String grpcHost = split.getGrpcHost().orElseThrow(() -> new PinotException(PINOT_INVALID_SEGMENT_QUERY_GENERATED, Optional.empty(), "Expected the segment split to contain the grpc host"));
```

Comment:
Is getGrpcPort guaranteed to be valid ... > 0 ?

Reply:
yes, if grpc port is < 0 then we will go to PinotSegmentPageSource instead of PinotSegmentStreamingPageSource

Reply:
Still, it's good to add a check here just in case.